File: backend/src/services/adminService.ts
import { supabaseAdmin } from '../config/supabase';

/**
 * Admin Service
 * Provides administrative analytics and user management functions
 */

export interface DashboardStats {
  totalUsers: number;
  activeUsers: number;
  totalPeople: number;
  totalRelationships: number;
  totalErrors: number;
  unresolvedErrors: number;
  errorRate24h: number;
  newUsersToday: number;
  newUsersThisWeek: number;
  avgPeoplePerUser: number;
}

export interface UserGrowthData {
  date: string;
  count: number;
}

export interface TreeSizeDistribution {
  size_range: string;
  user_count: number;
}

export interface ActiveUser {
  user_id: string;
  email: string;
  total_people: number;
  total_relationships: number;
  last_login_at: string | null;
  created_at: string;
}

export interface ErrorSummary {
  error_type: string;
  count: number;
  severity: string;
  latest_occurrence: string;
}

class AdminService {
  /**
   * Get dashboard statistics for admin overview
   */
  async getDashboardStats(): Promise<DashboardStats> {
    try {
      // Total users
      const { count: totalUsers } = await supabaseAdmin
        .from('user_metadata')
        .select('*', { count: 'exact', head: true });

      // Active users (logged in within last 30 days)
      const thirtyDaysAgo = new Date();
      thirtyDaysAgo.setDate(thirtyDaysAgo.getDate() - 30);
      const { count: activeUsers } = await supabaseAdmin
        .from('user_metadata')
        .select('*', { count: 'exact', head: true })
        .gte('last_login_at', thirtyDaysAgo.toISOString());

      // Total people records
      const { count: totalPeople } = await supabaseAdmin
        .from('persons')
        .select('*', { count: 'exact', head: true });

      // Total relationships
      const { count: totalRelationships } = await supabaseAdmin
        .from('relationships')
        .select('*', { count: 'exact', head: true });

      // Error statistics
      const { count: totalErrors } = await supabaseAdmin
        .from('error_logs')
        .select('*', { count: 'exact', head: true });

      const { count: unresolvedErrors } = await supabaseAdmin
        .from('error_logs')
        .select('*', { count: 'exact', head: true })
        .eq('resolved', false);

      // Error rate in last 24 hours
      const oneDayAgo = new Date();
      oneDayAgo.setHours(oneDayAgo.getHours() - 24);
      const { data: errorRate } = await supabaseAdmin
        .rpc('get_error_rate_by_hour', {
          hours_back: 24
        });
      const errorRate24h = errorRate?.[0]?.error_rate || 0;

      // New users today
      const today = new Date();
      today.setHours(0, 0, 0, 0);
      const { count: newUsersToday } = await supabaseAdmin
        .from('user_metadata')
        .select('*', { count: 'exact', head: true })
        .gte('created_at', today.toISOString());

      // New users this week
      const weekAgo = new Date();
      weekAgo.setDate(weekAgo.getDate() - 7);
      weekAgo.setHours(0, 0, 0, 0);
      const { count: newUsersThisWeek } = await supabaseAdmin
        .from('user_metadata')
        .select('*', { count: 'exact', head: true })
        .gte('created_at', weekAgo.toISOString());

      // Average people per user
      const avgPeoplePerUser = totalUsers && totalUsers > 0
        ? Math.round((totalPeople || 0) / totalUsers)
        : 0;

      return {
        totalUsers: totalUsers || 0,
        activeUsers: activeUsers || 0,
        totalPeople: totalPeople || 0,
        totalRelationships: totalRelationships || 0,
        totalErrors: totalErrors || 0,
        unresolvedErrors: unresolvedErrors || 0,
        errorRate24h,
        newUsersToday: newUsersToday || 0,
        newUsersThisWeek: newUsersThisWeek || 0,
        avgPeoplePerUser,
      };
    } catch (error) {
      console.error('Failed to get dashboard stats:', error);
      throw new Error('Failed to fetch dashboard statistics');
    }
  }

  /**
   * Get user growth data over time (daily for last 30 days)
   */
  async getUserGrowthData(days: number = 30): Promise<UserGrowthData[]> {
    try{
      const { data, error } = await supabaseAdmin
        .rpc('get_user_growth_by_day', { days_back: days });

      if (error) {
        console.error('Failed to get user growth data:', error);
        return [];
      }

      return data || [];
    } catch (error) {
      console.error('Failed to get user growth data:', error);
      return [];
    }
  }

  /**
   * Get tree size distribution (number of users by family tree size)
   */
  async getTreeSizeDistribution(): Promise<TreeSizeDistribution[]> {
    try {
      const { data, error } = await supabaseAdmin.rpc('get_tree_size_distribution');

      if (error) {
        console.error('Failed to get tree size distribution:', error);
        return [];
      }

      return data || [];
    } catch (error) {
      console.error('Failed to get tree size distribution:', error);
      return [];
    }
  }

  /**
   * Get most active users (by number of people created)
   */
  async getMostActiveUsers(limit: number = 10): Promise<ActiveUser[]> {
    try {
      // Get users with their people count
      const { data, error } = await supabaseAdmin
        .from('persons')
        .select('created_by_user_id')
        .order('created_at', { ascending: false });

      if (error) throw error;

      // Count people per user
      const userCounts = new Map<string, number>();
      data?.forEach(person => {
        if (person.created_by_user_id) {
          const count = userCounts.get(person.created_by_user_id) || 0;
          userCounts.set(person.created_by_user_id, count + 1);
        }
      });

      // Get top users
      const topUserIds = Array.from(userCounts.entries())
        .sort((a, b) => b[1] - a[1])
        .slice(0, limit)
        .map(([userId]) => userId);

      if (topUserIds.length === 0) return [];

      // Get user metadata for top users
      const { data: users, error: usersError } = await supabaseAdmin
        .from('user_metadata')
        .select('user_id, last_login_at, created_at')
        .in('user_id', topUserIds);

      if (usersError) throw usersError;

      // Combine with counts
      const activeUsers: ActiveUser[] = users?.map(user => ({
        user_id: user.user_id,
        email: '', // We don't have email in user_metadata, would need auth.users
        total_people: userCounts.get(user.user_id) || 0,
        total_relationships: 0, // TODO: Calculate relationships count
        last_login_at: user.last_login_at,
        created_at: user.created_at,
      })) || [];

      return activeUsers.sort((a, b) => b.total_people - a.total_people);
    } catch (error) {
      console.error('Failed to get active users:', error);
      return [];
    }
  }

  /**
   * Get error summary by type
   */
  async getErrorStats(days: number = 7): Promise<ErrorSummary[]> {
    try {
      const daysAgo = new Date();
      daysAgo.setDate(daysAgo.getDate() - days);

      const { data, error } = await supabaseAdmin
        .from('error_logs')
        .select('error_type, severity, timestamp')
        .gte('timestamp', daysAgo.toISOString())
        .order('timestamp', { ascending: false });

      if (error) throw error;

      // Group by error type
      const errorMap = new Map<string, ErrorSummary>();
      data?.forEach(err => {
        const key = err.error_type;
        if (!errorMap.has(key)) {
          errorMap.set(key, {
            error_type: err.error_type,
            count: 0,
            severity: err.severity,
            latest_occurrence: err.timestamp,
          });
        }
        const summary = errorMap.get(key)!;
        summary.count++;
        // Update to latest occurrence
        if (new Date(err.timestamp) > new Date(summary.latest_occurrence)) {
          summary.latest_occurrence = err.timestamp;
          summary.severity = err.severity; // Use severity from latest occurrence
        }
      });

      return Array.from(errorMap.values())
        .sort((a, b) => b.count - a.count);
    } catch (error) {
      console.error('Failed to get error stats:', error);
      return [];
    }
  }

  /**
   * Get system health metrics
   */
  async getSystemHealth() {
    try {
      // Test database connection
      const startTime = Date.now();
      const { error } = await supabaseAdmin
        .from('persons')
        .select('id', { count: 'exact', head: true })
        .limit(1);
      const dbLatency = Date.now() - startTime;

      return {
        status: error ? 'degraded' : 'healthy',
        database: {
          connected: !error,
          latency_ms: dbLatency,
        },
        timestamp: new Date().toISOString(),
      };
    } catch (error) {
      console.error('Failed to get system health:', error);
      return {
        status: 'unhealthy',
        database: {
          connected: false,
          latency_ms: -1,
        },
        timestamp: new Date().toISOString(),
      };
    }
  }

  /**
   * Update user role (super admin only)
   */
  async updateUserRole(userId: string, role: 'user' | 'admin' | 'super_admin'): Promise<void> {
    try {
      const { error } = await supabaseAdmin
        .from('user_metadata')
        .update({ role })
        .eq('user_id', userId);

      if (error) throw error;
    } catch (error) {
      console.error('Failed to update user role:', error);
      throw new Error('Failed to update user role');
    }
  }

  /**
   * Enable/disable user account
   */
  async updateUserStatus(userId: string, isActive: boolean): Promise<void> {
    try {
      const { error } = await supabaseAdmin
        .from('user_metadata')
        .update({ is_active: isActive })
        .eq('user_id', userId);

      if (error) throw error;
    } catch (error) {
      console.error('Failed to update user status:', error);
      throw new Error('Failed to update user status');
    }
  }

  /**
   * Mark error as resolved
   */
  async resolveError(errorId: string, adminUserId: string): Promise<void> {
    try {
      const { error } = await supabaseAdmin
        .from('error_logs')
        .update({
          resolved: true,
          resolved_at: new Date().toISOString(),
          resolved_by: adminUserId,
        })
        .eq('id', errorId);

      if (error) throw error;
    } catch (error) {
      console.error('Failed to resolve error:', error);
      throw new Error('Failed to resolve error');
    }
  }
}

export const adminService = new AdminService();

File: app/lib/services/admin_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../config/constants.dart';
import 'auth_service.dart';

final adminServiceProvider = Provider<AdminService>((ref) {
  final authService = ref.watch(authServiceProvider);
  return AdminService(authService);
});

class AdminService {
  final AuthService _authService;
  
  AdminService(this._authService);

  String get _baseUrl => AppConfig.apiBaseUrl;

  Map<String, String> get _headers => {
    'Content-Type': 'application/json',
    if (_authService.accessToken != null)
      'Authorization': 'Bearer ${_authService.accessToken}',
  };

  // ==================== DASHBOARD STATS ====================

  Future<DashboardStats> getDashboardStats() async {
    final response = await http.get(
      Uri.parse('$_baseUrl/admin/stats'),
      headers: _headers,
    );
    if (response.statusCode != 200) throw Exception('Failed to load stats');
    return DashboardStats.fromJson(jsonDecode(response.body));
  }

  // ==================== ANALYTICS ====================

  Future<List<UserGrowthData>> getUserGrowth({int days = 30}) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/admin/analytics/growth?days=$days'),
      headers: _headers,
    );
    if (response.statusCode != 200) throw Exception('Failed to load growth data');
    final List<dynamic> data = jsonDecode(response.body);
    return data.map((json) => UserGrowthData.fromJson(json)).toList();
  }

  Future<List<TreeSizeDistribution>> getTreeDistribution() async {
    final response = await http.get(
      Uri.parse('$_baseUrl/admin/analytics/tree-distribution'),
      headers: _headers,
    );
    if (response.statusCode != 200) throw Exception('Failed to load tree distribution');
    final List<dynamic> data = jsonDecode(response.body);
    return data.map((json) => TreeSizeDistribution.fromJson(json)).toList();
  }

  Future<List<ActiveUser>> getActiveUsers({int limit = 10}) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/admin/analytics/active-users?limit=$limit'),
      headers: _headers,
    );
    if (response.statusCode != 200) throw Exception('Failed to load active users');
    final List<dynamic> data = jsonDecode(response.body);
    return data.map((json) => ActiveUser.fromJson(json)).toList();
  }

  // ==================== ERROR LOGS ====================

  Future<ErrorLogsResponse> getErrorLogs({
    int page = 1,
    int pageSize = 50,
    String? type,
    String? severity,
  }) async {
    final queryParams = <String, String>{
      'page': page.toString(),
      'pageSize': pageSize.toString(),
      if (type != null) 'type': type,
      if (severity != null) 'severity': severity,
    };
    final uri = Uri.parse('$_baseUrl/admin/errors').replace(queryParameters: queryParams);
    final response = await http.get(uri, headers: _headers);
    if (response.statusCode != 200) throw Exception('Failed to load errors');
    return ErrorLogsResponse.fromJson(jsonDecode(response.body));
  }

  Future<List<ErrorSummary>> getErrorStats({int days = 7}) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/admin/errors/stats?days=$days'),
      headers: _headers,
    );
    if (response.statusCode != 200) throw Exception('Failed to load error stats');
    final List<dynamic> data = jsonDecode(response.body);
    return data.map((json) => ErrorSummary.fromJson(json)).toList();
  }

  Future<void> resolveError(String errorId) async {
    final response = await http.put(
      Uri.parse('$_baseUrl/admin/errors/$errorId/resolve'),
      headers: _headers,
    );
    if (response.statusCode != 200) throw Exception('Failed to resolve error');
  }

  // ==================== USER MANAGEMENT ====================

  Future<UsersResponse> getUsers({
    int page = 1,
    int pageSize = 20,
    String? role,
  }) async {
    final queryParams = <String, String>{
      'page': page.toString(),
      'pageSize': pageSize.toString(),
      if (role != null) 'role': role,
    };
    final uri = Uri.parse('$_baseUrl/admin/users').replace(queryParameters: queryParams);
    final response = await http.get(uri, headers: _headers);
    if (response.statusCode != 200) throw Exception('Failed to load users');
    return UsersResponse.fromJson(jsonDecode(response.body));
  }

  Future<void> updateUserRole(String userId, String role) async {
    final response = await http.put(
      Uri.parse('$_baseUrl/admin/users/$userId/role'),
      headers: _headers,
      body: jsonEncode({'role': role}),
    );
    if (response.statusCode != 200) throw Exception('Failed to update role');
  }

  Future<void> updateUserStatus(String userId, bool isActive) async {
    final response = await http.put(
      Uri.parse('$_baseUrl/admin/users/$userId/status'),
      headers: _headers,
      body: jsonEncode({'isActive': isActive}),
    );
    if (response.statusCode != 200) throw Exception('Failed to update status');
  }

  // ==================== AUDIT LOGS ====================

  Future<AuditLogsResponse> getAuditLogs({
    int page = 1,
    int pageSize = 50,
  }) async {
    final queryParams = <String, String>{
      'page': page.toString(),
      'pageSize': pageSize.toString(),
    };
    final uri = Uri.parse('$_baseUrl/admin/audit-logs').replace(queryParameters: queryParams);
    final response = await http.get(uri, headers: _headers);
    if (response.statusCode != 200) throw Exception('Failed to load audit logs');
    return AuditLogsResponse.fromJson(jsonDecode(response.body));
  }

  // ==================== SYSTEM HEALTH ====================

  Future<SystemHealth> getSystemHealth() async {
    final response = await http.get(
      Uri.parse('$_baseUrl/admin/health'),
      headers: _headers,
    );
    if (response.statusCode != 200) throw Exception('Failed to load health status');
    return SystemHealth.fromJson(jsonDecode(response.body));
  }
}

// ==================== MODELS ====================

class DashboardStats {
  final int totalUsers;
  final int activeUsers;
  final int totalPeople;
  final int totalRelationships;
  final int totalErrors;
  final int unresolvedErrors;
  final double errorRate24h;
  final int newUsersToday;
  final int newUsersThisWeek;
  final int avgPeoplePerUser;

  DashboardStats({
    required this.totalUsers,
    required this.activeUsers,
    required this.totalPeople,
    required this.totalRelationships,
    required this.totalErrors,
    required this.unresolvedErrors,
    required this.errorRate24h,
    required this.newUsersToday,
    required this.newUsersThisWeek,
    required this.avgPeoplePerUser,
  });

  factory DashboardStats.fromJson(Map<String, dynamic> json) => DashboardStats(
    totalUsers: json['totalUsers'] ?? 0,
    activeUsers: json['activeUsers'] ?? 0,
    totalPeople: json['totalPeople'] ?? 0,
    totalRelationships: json['totalRelationships'] ?? 0,
    totalErrors: json['totalErrors'] ?? 0,
    unresolvedErrors: json['unresolvedErrors'] ?? 0,
    errorRate24h: (json['errorRate24h'] ?? 0).toDouble(),
    newUsersToday: json['newUsersToday'] ?? 0,
    newUsersThisWeek: json['newUsersThisWeek'] ?? 0,
    avgPeoplePerUser: json['avgPeoplePerUser'] ?? 0,
  );
}

class UserGrowthData {
  final String date;
  final int count;

  UserGrowthData({required this.date, required this.count});

  factory UserGrowthData.fromJson(Map<String, dynamic> json) => UserGrowthData(
    date: json['date'],
    count: json['count'],
  );
}

class TreeSizeDistribution {
  final String sizeRange;
  final int userCount;

  TreeSizeDistribution({required this.sizeRange, required this.userCount});

  factory TreeSizeDistribution.fromJson(Map<String, dynamic> json) => TreeSizeDistribution(
    sizeRange: json['size_range'],
    userCount: json['user_count'],
  );
}

class ActiveUser {
  final String userId;
  final String email;
  final int totalPeople;
  final int totalRelationships;
  final String? lastLoginAt;
  final String createdAt;

  ActiveUser({
    required this.userId,
    required this.email,
    required this.totalPeople,
    required this.totalRelationships,
    this.lastLoginAt,
    required this.createdAt,
  });

  factory ActiveUser.fromJson(Map<String, dynamic> json) => ActiveUser(
    userId: json['user_id'],
    email: json['email'] ?? '',
    totalPeople: json['total_people'] ?? 0,
    totalRelationships: json['total_relationships'] ?? 0,
    lastLoginAt: json['last_login_at'],
    createdAt: json['created_at'],
  );
}

class ErrorLog {
  final String id;
  final DateTime timestamp;
  final String errorType;
  final String severity;
  final int? statusCode;
  final String message;
  final bool resolved;
  final DateTime? resolvedAt;

  ErrorLog({
    required this.id,
    required this.timestamp,
    required this.errorType,
    required this.severity,
    this.statusCode,
    required this.message,
    required this.resolved,
    this.resolvedAt,
  });

  factory ErrorLog.fromJson(Map<String, dynamic> json) => ErrorLog(
    id: json['id'],
    timestamp: DateTime.parse(json['error_timestamp']),
    errorType: json['error_type'],
    severity: json['severity'],
    statusCode: json['status_code'],
    message: json['message'],
    resolved: json['resolved'] ?? false,
    resolvedAt: json['resolved_at'] != null ? DateTime.parse(json['resolved_at']) : null,
  );
}

class ErrorLogsResponse {
  final List<ErrorLog> errors;
  final Pagination pagination;

  ErrorLogsResponse({required this.errors, required this.pagination});

  factory ErrorLogsResponse.fromJson(Map<String, dynamic> json) => ErrorLogsResponse(
    errors: (json['errors'] as List).map((e) => ErrorLog.fromJson(e)).toList(),
    pagination: Pagination.fromJson(json['pagination']),
  );
}

class ErrorSummary {
  final String errorType;
  final int count;
  final String severity;
  final String latestOccurrence;

  ErrorSummary({
    required this.errorType,
    required this.count,
    required this.severity,
    required this.latestOccurrence,
  });

  factory ErrorSummary.fromJson(Map<String, dynamic> json) => ErrorSummary(
    errorType: json['error_type'],
    count: json['count'],
    severity: json['severity'],
    latestOccurrence: json['latest_occurrence'],
  );
}

class UserMetadata {
  final String userId;
  final String role;
  final bool isActive;
  final DateTime? lastLoginAt;
  final DateTime createdAt;

  UserMetadata({
    required this.userId,
    required this.role,
    required this.isActive,
    this.lastLoginAt,
    required this.createdAt,
  });

  factory UserMetadata.fromJson(Map<String, dynamic> json) => UserMetadata(
    userId: json['user_id'],
    role: json['role'],
    isActive: json['is_active'] ?? true,
    lastLoginAt: json['last_login_at'] != null ? DateTime.parse(json['last_login_at']) : null,
    createdAt: DateTime.parse(json['created_at']),
  );
}

class UsersResponse {
  final List<UserMetadata> users;
  final Pagination pagination;

  UsersResponse({required this.users, required this.pagination});

  factory UsersResponse.fromJson(Map<String, dynamic> json) => UsersResponse(
    users: (json['users'] as List).map((e) => UserMetadata.fromJson(e)).toList(),
    pagination: Pagination.fromJson(json['pagination']),
  );
}

class AuditLog {
  final String id;
  final String adminUserId;
  final String actionType;
  final String resourceType;
  final String? resourceId;
  final DateTime timestamp;

  AuditLog({
    required this.id,
    required this.adminUserId,
    required this.actionType,
    required this.resourceType,
    this.resourceId,
    required this.timestamp,
  });

  factory AuditLog.fromJson(Map<String, dynamic> json) => AuditLog(
    id: json['id'],
    adminUserId: json['admin_user_id'],
    actionType: json['action_type'],
    resourceType: json['resource_type'],
    resourceId: json['resource_id'],
    timestamp: DateTime.parse(json['timestamp']),
  );
}

class AuditLogsResponse {
  final List<AuditLog> logs;
  final Pagination pagination;

  AuditLogsResponse({required this.logs, required this.pagination});

  factory AuditLogsResponse.fromJson(Map<String, dynamic> json) => AuditLogsResponse(
    logs: (json['logs'] as List).map((e) => AuditLog.fromJson(e)).toList(),
    pagination: Pagination.fromJson(json['pagination']),
  );
}

class SystemHealth {
  final String status;
  final Map<String, dynamic> database;
  final String timestamp;

  SystemHealth({
    required this.status,
    required this.database,
    required this.timestamp,
  });

  factory SystemHealth.fromJson(Map<String, dynamic> json) => SystemHealth(
    status: json['status'],
    database: json['database'],
    timestamp: json['timestamp'],
  );
}

class Pagination {
  final int page;
  final int pageSize;
  final int total;
  final int totalPages;

  Pagination({
    required this.page,
    required this.pageSize,
    required this.total,
    required this.totalPages,
  });

  factory Pagination.fromJson(Map<String, dynamic> json) => Pagination(
    page: json['page'],
    pageSize: json['pageSize'],
    total: json['total'],
    totalPages: json['totalPages'],
  );
}
